### ISSUE TITLE
Support different partitions under the same multi asset

### ISSUE THREAD

**Issue by askvinni** [dagster-io/dagster#13616]
### What's the use case?

For a current use case, I'm using a `multi_asset` to generate n assets with a `MultiPartitionsDefinition`, some of which should should have both partition sets, and some only one. Simplified, it looks like the following:
```
@multi_asset(
    outs={
        "some_ab_asset": AssetOut(), # should include both a and b as static partitions and all dynamic
        "a_asset": AssetOut(is_required=False), # should either only include dynamic partitions, or only "a" and all dynamic
        "b_asset": AssetOut(is_required=False), # should either only include dynamic partitions, or only "b" and all dynamic
    },
    partitions_def=MultiPartitionsDefinition(
        {
            "static": StaticPartitionsDefinition(["a", "b"]),
            "dynamic": DynamicPartitionsDefinition(name="my_dynamic_partition"),
        }
    ),
    can_subset=True,
)
```
Currently the workaround is to just return None if the asset is not expected to exist for the static partition called, but it would be great for observability if dagster could know that some of the assets need to be partitioned differently.

### Ideas of implementation

Ergonomically, I would naturally drift towards specifying the partition sets in the `AssetOut` function.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by clairelin135** [comment]
I don't think there are any blockers for this at the moment, we would just need to thread some `partitions_def_by_output_asset_key` arg throughout the system and do testing for this.

**Comment by askvinni** [comment]
@clairelin135 that argument would also be a good ergonomic choice! One thing I thought of now is that depending on how the partitions are passed to the `multi_asset` function, it could cause some finicky errors. If one or more of the underlying assets require a `MultiPartitionsDefinition` while others have one-dimensional partitioning like my example, some care might be necessary in passing/fetching `context.partition_key.keys_by_dimension`.

**Comment by sryza** [comment]
A difficulty here is that Dagster currently has a constraint that each run targets a single partition or single range of partitions.  I.e. all the partitioned assets in the run must materialize that partition. If I understand correctly, we'd need to break that constraint for this to work?

**Comment by askvinni** [comment]
@sryza you're right, this would mean effectively targeting different partitions in a single run. Although I do remember seeing somewhere on slack that you had plans to allow mapping outputs to dynamic partitions for an originally unpartitioned run (or did I misunderstand that)? If so, I'd suppose that that constraint would be lifted, and I'd be happy to put something together for this case when it is.
